metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/dao/WorkflowValidationUtils.java

```diff
@@ -0,0 +1,216 @@
+    validateAndTrimHarvestParameters(workflow.getDatasetId(), enabledPlugins);
+
+    // Check that first plugin is not link checking (except if it is the only plugin)
+    if (enabledPlugins.size() > 1
+            && enabledPlugins.get(0).getPluginType() == PluginType.LINK_CHECKING) {
+      throw new PluginExecutionNotAllowed(CommonStringValues.PLUGIN_EXECUTION_NOT_ALLOWED);
+    }
+
+    // Make sure that all enabled plugins (except the first) have a predecessor within the workflow.
+    final EnumSet<ExecutablePluginType> previousTypesInWorkflow = EnumSet
+            .of(enabledPlugins.get(0).getExecutablePluginType());
+    for (int i = 1; i < enabledPlugins.size(); i++) {
+
+      // Find the permissible predecessors
+      final ExecutablePluginType pluginType = enabledPlugins.get(i).getExecutablePluginType();
+      final Set<ExecutablePluginType> permissiblePredecessors = DataEvolutionUtils
+              .getPredecessorTypes(pluginType);
+
+      // Check if we have the right predecessor plugin types in the workflow
+      final boolean hasNoPredecessor = !permissiblePredecessors.isEmpty() &&
+              permissiblePredecessors.stream().noneMatch(previousTypesInWorkflow::contains);
+      if (hasNoPredecessor) {
+        throw new PluginExecutionNotAllowed(CommonStringValues.PLUGIN_EXECUTION_NOT_ALLOWED);
+      }
+
+      // Add the plugin type to those we have seen
+      previousTypesInWorkflow.add(pluginType);
+    }
+
+    // We should now have seen all types. Make sure that there are no duplicates
+    if (previousTypesInWorkflow.size() != enabledPlugins.size()) {
+      throw new PluginExecutionNotAllowed(CommonStringValues.PLUGIN_EXECUTION_NOT_ALLOWED);
+    }
+
+    // Check the presence of the predecessor and return it.
+    return dataEvolutionUtils
+            .computePredecessorPlugin(enabledPlugins.get(0).getExecutablePluginType(),
+                    enforcedPredecessorType, workflow.getDatasetId());
+  }
+
+  private void validateAndTrimHarvestParameters(String datasetId,
+          Iterable<AbstractExecutablePluginMetadata> enabledPlugins) throws BadContentException {
+    for (AbstractExecutablePluginMetadata pluginMetadata : enabledPlugins) {
+      if (pluginMetadata instanceof OaipmhHarvestPluginMetadata) {
+        final OaipmhHarvestPluginMetadata oaipmhMetadata = (OaipmhHarvestPluginMetadata) pluginMetadata;
+        final URI validatedUri = validateUrl(oaipmhMetadata.getUrl());
+        oaipmhMetadata
+                .setUrl(new URIBuilder(validatedUri).removeQuery().setFragment(null).toString());
+        oaipmhMetadata.setMetadataFormat(oaipmhMetadata.getMetadataFormat() == null ? null
+                : oaipmhMetadata.getMetadataFormat().trim());
+        oaipmhMetadata.setSetSpec(
+                oaipmhMetadata.getSetSpec() == null ? null : oaipmhMetadata.getSetSpec().trim());
+        if (oaipmhMetadata.isIncrementalHarvest() && !isIncrementalHarvestingAllowed(datasetId)) {
```

Comment:
We should capture this case in a unit test

Reply:
Yes, I can add this.